internal/telemetry/metric_exporter.go

```diff
@@ -0,0 +1,341 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/aggregation"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+
+	"github.com/elastic/apm-data/model/modelpb"
+)
+
+// NewMetricExporter initializes a new MetricExporter
+func NewMetricExporter(p modelpb.BatchProcessor, opts ...ConfigOption) *MetricExporter {
+	cfg := newConfig(opts...)
+
+	return &MetricExporter{
+		processor: p,
+
+		metricFilter:        cfg.MetricFilter,
+		temporalitySelector: cfg.TemporalitySelector,
+		aggregationSelector: cfg.AggregationSelector,
+	}
+}
+
+// MetricExporter is an OpenTelemetry metric Reader which retrieves metrics,
+// filters them and emits them to the specified consumer
+type MetricExporter struct {
+	processor modelpb.BatchProcessor
+
+	metricFilter        []string
+	temporalitySelector metric.TemporalitySelector
+	aggregationSelector metric.AggregationSelector
+}
+
+// Temporality returns the Temporality to use for an instrument kind.
+func (e *MetricExporter) Temporality(k metric.InstrumentKind) metricdata.Temporality {
+	return e.temporalitySelector(k)
+}
+
+// Aggregation returns the Aggregation to use for an instrument kind.
+func (e *MetricExporter) Aggregation(k metric.InstrumentKind) aggregation.Aggregation {
+	return e.aggregationSelector(k)
+}
+
+func (e *MetricExporter) Export(ctx context.Context, rm *metricdata.ResourceMetrics) error {
+	batch := modelpb.Batch{}
+	now := time.Now()
+
+	baseEvent := modelpb.APMEvent{
+		Service: &modelpb.Service{
+			Name:     "apm-server",
+			Language: &modelpb.Language{Name: "go"},
+		},
+		Agent: &modelpb.Agent{
+			Name:    "internal",
+			Version: "unknown",
+		},
```

Comment:
Is this needed?

Reply:
apm-data sets it, I figured kibana may uses it somehow?

Reply:
Yes, under the assumption that all metrics are coming from an external agent. I'm not sure it makes sense here. These metrics won't show up in the APM app, so I think we can remove these fields.

Reply:
I have removed the agent. I've kept the service, as we'd still want to know where the metric comes from.

Reply:
:+1: thanks